### PR TITLE
[Enhancement] Support sortkey reorder

### PR DESCRIPTION
**Pull request by TszKitLo40**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|20|20|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|20|20|100.00%|[]|

**Review by luohaha — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|18|20|90.00%|[335, 336]|

File: be/src/storage/lake/schema_change.cpp [hunk 1]
```diff
@@ -318,6 +318,37 @@ Status SchemaChangeHandler::do_process_alter_tablet(const TAlterTabletReqV2& req
                                                      sc_params.materialized_params_map, has_delete_predicates,
                                                      &sc_params.sc_sorting, &sc_params.sc_directly, nullptr));
 
+    if (base_schema->keys_type() == KeysType::PRIMARY_KEYS) {
```
Comment (wyb) [inline be/src/storage/lake/schema_change.cpp:321]: better to put this in `SchemaChangeUtils::parse_request`

File: be/test/storage/lake/schema_change_test.cpp [hunk 2]
```diff
@@ -877,4 +877,242 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+        CHECK_OK(fs::create_directories(lake::join_path(kTestGroupPath, lake::kMetadataDirectoryName)));
+        CHECK_OK(fs::create_directories(lake::join_path(kTestGroupPath, lake::kTxnLogDirectoryName)));
+        CHECK_OK(_tablet_manager->put_tablet_metadata(*_base_tablet_metadata));
+        CHECK_OK(_tablet_manager->put_tablet_metadata(*_new_tablet_metadata));
+    }
+
+    void TearDown() override { (void)fs::remove_all(kTestGroupPath); }
+};
+
+TEST_P(SchemaChangeSortKeyReorderTest, test_alter_sortkey_reorder) {
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1003]: Test the case where the new order key is not a subset of the old order key

File: be/test/storage/lake/schema_change_test.cpp [hunk 3]
```diff
@@ -877,4 +877,242 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(SchemaChangeSortKeyReorderTest, SchemaChangeSortKeyReorderTest,
+                         ::testing::Values(SchemaChangeParam{PRIMARY_KEYS, 0, 0},
+                                           SchemaChangeParam{PRIMARY_KEYS, 0, 1},
+                                           SchemaChangeParam{PRIMARY_KEYS, 0, 2},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 0},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 1},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 2},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 4}),
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1114]: Maybe it's not necessary to test so many different data sizes, they don't make much sense for schema change.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/test/storage/lake/schema_change_test.cpp [hunk 4]
```diff
@@ -885,4 +885,476 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+        // new tablet
+        _new_tablet_metadata = std::make_shared<TabletMetadata>();
+        _new_tablet_metadata->set_id(next_id());
+        _new_tablet_metadata->set_version(1);
+        //
+        //  | column | type | KEY | NULL | SORTKEY |
+        //  +--------+------+-----+------+---------+
+        //  |   c0   |  INT | YES |  NO  |   NO    |
+        //  |   c1   |  INT | NO  |  NO  |   YES   |
+        //  |   c2   |  INT | NO  |  NO  |   NO    |
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1188]:
```suggestion
        //  | column | type | KEY | NULL | SORTKEY |
        //  +--------+------+-----+------+---------+
        //  |   c0   |  INT | YES |  NO  |   YES    |
        //  |   c1   |  INT | NO  |  NO  |   YES   |
        //  |   c2   |  INT | NO  |  NO  |   YES    |
```

File: be/test/storage/lake/schema_change_test.cpp [hunk 5]
```diff
@@ -885,4 +885,476 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:953]:
```suggestion
        //  | column | type | KEY | NULL | SORTKEY |
        //  +--------+------+-----+------+---------+
        //  |   c0   |  INT | YES |  NO  |   NO    |
        //  |   c1   |  INT | NO  |  NO  |   NO   |
        //  |   c2   |  INT | NO  |  NO  |   YES    |
```

File: be/test/storage/lake/schema_change_test.cpp [hunk 6]
```diff
@@ -885,4 +885,476 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+    cv0->append_numbers(v0.data(), v0.size() * sizeof(int));
+
+    VChunk chunk0({ck0, ck1, cv0}, _base_schema);
+
+    auto indexes = std::vector<uint32_t>(k0.size());
+    for (int i = 0; i < k0.size(); i++) {
+        indexes[i] = i;
+    }
+
+    // write 2 rowsets
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1267]:
```suggestion
```

File: be/test/storage/lake/schema_change_test.cpp [hunk 7]
```diff
@@ -885,4 +885,474 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(SchemaChangeSortKeyReorderTest2, SchemaChangeSortKeyReorderTest2,
+                         ::testing::Values(SchemaChangeParam{PRIMARY_KEYS, 0, 1},
+                                           SchemaChangeParam{PRIMARY_KEYS, 0, 2},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 1},
+                                           SchemaChangeParam{PRIMARY_KEYS, 2, 4}),
+                         to_string_param_name);
+// clang-format on
+
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1357]: Add a test case that the original sort keys are c1, c2 and the new sort key is c1?

File: be/test/storage/lake/schema_change_test.cpp [hunk 8]
```diff
@@ -885,4 +885,474 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
+    VChunk chunk0({ck0, ck1, cv0}, _base_schema);
+
+    auto indexes = std::vector<uint32_t>(k0.size());
+    for (int i = 0; i < k0.size(); i++) {
+        indexes[i] = i;
+    }
+
+    int64_t version = 1;
+    int64_t txn_id = 1000;
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1031]:
```suggestion
    int64_t txn_id = next_id();
```

File: be/test/storage/lake/schema_change_test.cpp [hunk 9]
```diff
@@ -885,4 +885,474 @@ INSTANTIATE_TEST_SUITE_P(SchemaChangeModifyColumnOrderTest, SchemaChangeModifyCo
...
```
Comment (sduzh) [inline be/test/storage/lake/schema_change_test.cpp:1267]:
```suggestion
    int64_t txn_id = next_id();
```

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/lake/schema_change.cpp|18|20|90.00%|[335, 336]|

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/schema_change_utils.cpp|20|20|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/schema_change_utils.cpp|20|20|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/schema_change_utils.cpp|20|20|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**